### PR TITLE
ws-manager: maintenance mode

### LINKED ISSUE #15409: [ws-manager] maintenance mode
## Is your feature request related to a problem? Please describe

We sometimes want to upgrade workspace component images in existing workspace clusters, such as `registry-facade`, `ws-daemon`, and `ws-manager`, but cannot because they are starting workspaces (pulling images, initializing content), stopping (backing content up to object storage) or taking a snapshot (backing up too).

## Describe the behaviour you'd like

1. Persist that maintenance mode is enabled in the `ws-manager` configmap
   * `maintenanceMode` should be persisted as a new object that for now only has an `enabled` field
   * Maintenance mode should only be considered active when enabled is true
2. `ws-manager` must reload the config when maintenanceMode is altered in the configmap
~3. Add two new gRPC endpoints to ws-manager: startMaintenance, stopMaintenance~
   ~* Start sets enabled to true in the configmap~
   ~* Stop sets enabled to false in the configmap~
4. While in maintenance mode, the following gRPC endpoints must fail in a state that is predictable
   * StartWorkspace
   * StopWorkspace
   * TakeSnapshot
5. While in maintenance mode workspace timeouts (which trigger backup) must be delayed until maintenanceMode is disabled

## Describe alternatives you've considered

All other [wsmanager gRPC](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/core.proto) do not require changes and should fail gracefully w/o alternation to be considered for maintenance mode.

## Additional context

Maintenance mode will not be used with gitpod.io. It will be used with Gitpod Dedicated, as a short term solution to do in-place upgrades, as we do not yet have workspace mk2.

## Questions
How will this impact related metrics for workspace start and stop, with and without failures?

### PR DESCRIPTION
**Pull request by jenting**
## Description

A short-term solution for upgrading the ws-manager within a cluster rather than traffic shift to a new cluster.
Before upgrading the ws-manager to a newer one, we will
- update the ws-manager ConfigMap `"maintenanceMode": true,`.
- the other component will check if is there any workspace backing content to remote _or_ takes a snapshot. If none, replace the ws-manager with a newer one.
- after rolling out the ws-manager finished, update the ws-manager ConfigMap `"maintenanceMode": false,`.

While the maintenance mode is enabled, the user can't start the workspace, stop the workspace, or take snapshots. Also, during the maintenance mode, the workspace would not time out.

Suppose there is workspace time out during the maintenance mode. After maintenance mode is disabled, the ws-manager needs to restart or restart a new one to stop the timed-out workspace.

## Related Issue(s)

Fixes #15409

## How to test

Test it by gRPC call
```console
# install grpcurl
go install github.com/fullstorydev/grpcurl/cmd/grpcurl@latest

# port-forward ws-manager port 8080 locally
kubectl port-forward deployment/ws-manager 8080:8080

# update the ws-manager ConfigMap, add `"maintenanceMode": true,`
# waits for the configuration hot-reloading (check the ws-manager log with `reloading file after change`)
kubectl edit cm ws-manager

# run grpcurl, make sure the below requests response status code `FailedPrecondition` and error message `under maintenance mode`.
grpcurl -insecure -proto /workspace/gitpod/components/ws-manager-api/core.proto -import-path=/workspace/gitpod/components/ localhost:8080 wsman.WorkspaceManager.StartWorkspace
grpcurl -insecure -proto /workspace/gitpod/components/ws-manager-api/core.proto -import-path=/workspace/gitpod/components/ localhost:8080 wsman.WorkspaceManager.StopWorkspace
grpcurl -insecure -proto /workspace/gitpod/components/ws-manager-api/core.proto -import-path=/workspace/gitpod/components/ localhost:8080 wsman.WorkspaceManager.TakeSnapshot
```

User experience when maintenanceMode is enabled.
- Start Workspace
   image
- Stop Workspace: nothing happens
- Take Snapshot
   image

## Release Notes

```release-note
None
```

## Documentation

None

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/ws-manager/pkg/manager/manager.go
Comment (kylos101) [inline components/ws-manager/pkg/manager/manager.go:819]: What was dry run for? Just curious. :smile:

File: components/ws-manager/pkg/manager/manager.go
Comment (kylos101) [inline components/ws-manager/pkg/manager/manager.go:820]: Should this message return under maintenance mode, too, rather than dry run message?

**Review by kylos101 — COMMENTED** [review]
Just some drive by questions :car: , cool to see this simplified and just rely on the config change.

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [reply components/ws-manager/pkg/manager/manager.go:819]: That's a good question. To be honest I don't know.
But I checked the related code and did not find anywhere using the dryRun.

I did consider reusing the `dryRun` term to implement the maintenance mode.
However, it makes us confused that dryRun equals to maintenance mode.

Therefore, deprecating the dryRun and adding a new maintenanceMode configuration is reasonable.

**Comment by jenting** [comment]
With the current code change. During the maintenance mode is enabled, and a workspace is timed out. 

The good news is we would not stop the workspace and trigger backing up content to the remote.
The bad news is after maintenance mode is disabled, this workspace will not time out anymore.

The problem is there is no more event coming for this workspace, so the ws-manager can't stop this workspace.

Is there any way to keep reconciling this workspace pod? cc @csweichel

**Comment by jenting** [comment]
@csweichel PTAL

**Comment by kylos101** [comment]
> @csweichel PTAL

@jenting did you forget to push a commit? I ask because @csweichel shared an idea [here](https://github.com/gitpod-io/gitpod/pull/15444#issuecomment-1361290017) to help with your timeout question, but, I do not see a related commit.

**Comment by jenting** [comment]
> > @csweichel PTAL
> 
> @jenting did you forget to push a commit? I ask because @csweichel shared an idea [here](https://github.com/gitpod-io/gitpod/pull/15444#issuecomment-1361290017) to help with your timeout question, but, I do not see a related commit.

No. Restarting the ws-manager should rely on the external component. After the maintenance mode is disabled, restart the ws-manager (restart the newer ws-manager version).

**Comment by kylos101** [comment]
@jenting I just rebased with main and did a force push because, you'll :laughing: , my preview environment was unexpectedly purged while testing!

**Comment by kylos101** [comment]
@jenting 
suggestion: The description lacks some use cases, consider adding them.
1. a workspace pod is starting and pulling an image through registry-facade (regular or prebuild).
2. an imagebuild pod is running, potentially pulling through registry-facade

In both cases, the external component should wait for image pull scenarios to finish, in addition to the content init and workspace backup scenarios.

**Comment by kylos101** [comment]
I cannot stop workspaces. :+1: 

I'm able to expose/unexpose ports given `curl lama.sh | sh`, which is expected seeing that is not governed by maintenanceMode. :+1: 

TakeSnapshot fails with `Cannot capture workspace snapshot: Error: Request takeSnapshot failed with message: Channel has been shut down`. :+1:

**Comment by kylos101** [comment]
Recreating the VM because fluentbit is crashlooping and I'd like to examine logs inbetween restarts.

**Comment by kylos101** [comment]
/werft run with-large-vm=true with-preview=true recreate-vm=true

**Review by kylos101 — APPROVED** [review]
Solid work, @jenting! 

Below are my test notes.

With maintenanceMode enabled

I cannot start regular workspaces and receive `13 INTERNAL: cannot start build: "rpc error: code = FailedPrecondition desc = under maintenance mode"`. :+1: 

I cannot stop workspaces via the UI or `gp stop`. :+1: There is no feedback from `gp stop` when this is done, like there is for `gp snapshot`.

I'm able to expose/unexpose ports given `curl lama.sh | sh`, which is expected seeing that is not governed by maintenanceMode. :+1: 

TakeSnapshot and `gp snapshot` fail with `jsonrpc2: code -32603 message: Request takeSnapshot failed with message: 9 FAILED_PRECONDITION: under maintenance mode`. :+1: 

prebuilds fail with `Prebuild failed for system reasons. Please contact support. Error: 13 INTERNAL: cannot start build: "rpc error: code = FailedPrecondition desc = under maintenance mode"`

imagebuilds get `13 INTERNAL: cannot start build: "rpc error: code = FailedPrecondition desc = under maintenance mode"`.

Also, regarding metrics:
- starts_total does not increment
- stops_total does not increment
- take snapshot has no metric in ws-manager to check, as its a grpc call to ws-daemon, and we never get that far

With maintenanceMode disabled:

I was able to stop workspaces that were running during the maintenance window. I was able to start new workspaces, too.

**Comment by kylos101** [comment]
@aledbf @sagor999 we require one of your approvals, too, as code owners (I assume of `ws-manager-api`). :pray:

File: components/ws-manager/pkg/manager/manager.go
Comment (aledbf) [inline components/ws-manager/pkg/manager/manager.go:819]: What does this mean for users? Can they not stop workspaces?

Comment (kylos101) [reply components/ws-manager/pkg/manager/manager.go:819]: Yes, they will not be able to stop until maintenanceMode is disabled, including timeouts that would have elapsed during the maintenance window.

Comment (kylos101) [reply components/ws-manager/pkg/manager/manager.go:819]: @aledbf let @jenting or I know if you need anything else? :pray:

File: components/ws-manager-api/go/config/config.go
Comment (sagor999) [inline components/ws-manager-api/go/config/config.go:113]:
```suggestion
	// MaintenanceMode prevents start workspace, stop workspace, and take snapshot operations
```

File: components/ws-manager/cmd/run.go
Comment (sagor999) [inline components/ws-manager/cmd/run.go:127]:
```suggestion
			log.WithError(err).WithField("cfgFile", cfgFile).Fatal("cannot start watch of configuration file")
```
to know where it is looking for config

File: components/ws-manager/pkg/manager/manager.go
Comment (sagor999) [reply components/ws-manager/pkg/manager/manager.go:819]: what happens to user's credits if they cannot stop workspace? :thinking:

**Review by sagor999 — APPROVED** [review]
/hold

nits and comments

File: components/ws-manager/pkg/manager/manager.go
Comment (kylos101) [reply components/ws-manager/pkg/manager/manager.go:819]: Good question @sagor999 . I believe we'll continue to burn through the credits when using this maintenance mode.

But, I think that's a side effect of not having mk2, and usage of this maintenance mode will be limited to Gitpod Dedicated for now. 

In other words, once we have mk2, I assume we'd persist the stop time request within the CR rather than return the failed precondition, and ideally leverage the time of the stop request to avoid burning through more credits.

Comment (jenting) [reply components/ws-manager/pkg/manager/manager.go:819]: @aledbf @sagor999 Are there any other questions/concerns? If not, please help unhold it. Thank you.

**Pull request merged**
